x-pack/plugins/aiops/server/routes/log_rate_analysis/analysis_handlers/top_items_handler.ts

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { queue } from 'async';
+
+import { SIGNIFICANT_ITEM_TYPE, type SignificantItem } from '@kbn/ml-agg-utils';
+import { i18n } from '@kbn/i18n';
+
+import {
+  addSignificantItemsAction,
+  updateLoadingStateAction,
+} from '../../../../common/api/log_rate_analysis/actions';
+
+import { isRequestAbortedError } from '../../../lib/is_request_aborted_error';
+
+import { fetchTopCategories } from '../queries/fetch_top_categories';
+import { fetchTopTerms } from '../queries/fetch_top_terms';
+
+import type {
+  AiopsLogRateAnalysisSchema,
+  AiopsLogRateAnalysisApiVersion as ApiVersion,
+} from '../../../../common/api/log_rate_analysis/schema';
+
+import {
+  LOADED_FIELD_CANDIDATES,
+  MAX_CONCURRENT_QUERIES,
+  PROGRESS_STEP_P_VALUES,
+} from '../response_stream_utils/constants';
+import type { ResponseStreamFetchOptions } from '../response_stream_factory';
+
+export const topItemsHandlerFactory =
+  <T extends ApiVersion>({
+    abortSignal,
+    client,
+    logDebugMessage,
+    logger,
+    requestBody,
+    responseStream,
+    stateHandler,
+    version,
+  }: ResponseStreamFetchOptions<T>) =>
+  async ({
+    fieldCandidates,
+    textFieldCandidates,
+  }: {
+    fieldCandidates: string[];
+    textFieldCandidates: string[];
+  }) => {
+    let fieldCandidatesCount = fieldCandidates.length;
+
+    // This will store the combined count of detected log patterns and keywords
+    let fieldValuePairsCount = 0;
+
+    const topCategories: SignificantItem[] = [];
+
+    if (version === '1') {
+      topCategories.push(
+        ...((requestBody as AiopsLogRateAnalysisSchema<'1'>).overrides?.significantTerms?.filter(
+          (d) => d.type === SIGNIFICANT_ITEM_TYPE.LOG_PATTERN
+        ) ?? [])
+      );
+    }
+
+    if (version === '2') {
```

Comment:
Version 1 and 2 look pretty much identical here - is this duplication purely for type reasons?

Reply:
Just a small difference, version 1 is accessing `significantTerms` whereas version 2 uses `significantItems`.